[Infra] Adding a button to snapshot the current Media Browse Tree

Leo-Neat opened the pull request:
The MCT app already has a section that lists the single level contents of a media browse tree. This section can be accessed by clicking the control button then swiping to the media  items tab. This page displays the list of media items at the current level of the media browse tree. There are two buttons, up and top. The Up button brings you one level higher on the browse tree and the top button takes you to the root node. This page is useful but only gives us a single level of the tree and not the whole tree itself. We propose to add a button next to the “Top” and “Up” that will be labeled “Save to File”.
This button will start at the root node and run a depth first search tree traversal printing out each of the browse tree elements in the following format. Each subsequent level will have an added tab character in front of the media items description. One possible output could look like this:

nic0lette commented on mediacontroller/src/main/java/com/example/android/mediacontroller/MediaAppControllerActivity.java:
```diff
@@ -1228,6 +1291,173 @@ void init(View topButtonView, View upButtonView) {
+                }
+                subscribe();
+                saveButtonView.setOnClickListener(
+                        v -> {
```
Please refactor all of this logic out to a separate file that (ideally) would handle all of this separately from the UI (to avoid the race condition at all).

It would also be really great if the new code were in Kotlin.

Leo-Neat replied:
Done.

nic0lette commented on mediacontroller/src/main/AndroidManifest.xml:
```diff
@@ -17,6 +17,8 @@
     xmlns:tools="http://schemas.android.com/tools"
     package="com.example.android.mediacontroller">
 
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
```
I mentioned this in #27 but this will not work on API 30+, so I'd request that you do not use this permission and hard code a path, but instead either:

1) Use the intent `ACTION_CREATE_DOCUMENT` to allow the user to pick a place to put the file
2) Put the file in the "Downloads" folder/media store collection.

If you choose to go with option 2, then you can add `maxSdkVersion=29` to this permission and not worry about requesting it on API 30.

Leo-Neat replied:
Fixed, used ACTION_CREATE_DOCUMENT.

nic0lette commented on mediacontroller/src/main/java/com/example/android/mediacontroller/MediaAppControllerActivity.java:
```diff
@@ -1228,6 +1291,173 @@ void init(View topButtonView, View upButtonView) {
+                            }
+
+                            // Create output file
+                            File root = android.os.Environment.getExternalStorageDirectory();
```
This is, in particular, the part that doesn't work on API 30, as apps don't get write access to the root of the sdcard, and should prefer saving files in Downloads or Documents.

Leo-Neat replied:
Fixed.

nic0lette commented on mediacontroller/src/main/java/com/example/android/mediacontroller/MediaAppControllerActivity.java:
```diff
@@ -1206,7 +1268,8 @@ void updateItems(List<MediaBrowserCompat.MediaItem> items) {
          * for moving up one level in the tree.
          */
-        void init(View topButtonView, View upButtonView) {
+        @RequiresApi(api = Build.VERSION_CODES.N)
```
This doesn't look right? If `init()` requires Android N+, doesn't that mean the entire thing would require N+?

Leo-Neat replied:
Fixed.

nic0lette reviewed (changes requested):
I added some comments, but overall I think this really most of this needs to be split into a separate file and more, smaller methods. A method for doing the DFS, a method for printing a media item, etc...

I'm also not sure why the method requires Android N+? In generally I'd really like to keep the minSdkVersion=21 (I'd actually be fine raising it to 23, for various reasons, but not 24 yet). So I'd also prefer if you could guard the limited things that actually need Android N+ with a `Build.VERSION.SDK_INT` check around those bits, rather than annotating the entire method.

nic0lette commented on mediacontroller/src/main/java/com/example/android/mediacontroller/MediaBrowseTreeSnapshot.kt:
```diff
@@ -0,0 +1,170 @@
+import java.util.concurrent.Executors
+import java.util.concurrent.Semaphore
+
+class MediaBrowseTreeSnapshot(private val mBrowser: MediaBrowserCompat, private val mContext: Context) {
```
Kotlin doesn't use 'm' prefixes in parameters or member variables.

nit: Ideally `context` should be the first parameter

Leo-Neat replied:
Done.

nic0lette commented on mediacontroller/src/main/java/com/example/android/mediacontroller/MediaBrowseTreeSnapshot.kt:
```diff
@@ -0,0 +1,170 @@
+        val loaded = Semaphore(1)
+        val executorService = Executors.newFixedThreadPool(4)
+        val mItems: MutableList<MediaBrowserCompat.MediaItem> = ArrayList()
+        executorService.execute {
```
Since this is in Kotlin, you should use suspend functions, not executors, to do the background work.

Leo-Neat replied:
Done.

nic0lette reviewed (commented):
I was taking a quick look and wanted to provide this feedback early since it may require digging into suspend functions a bit more than you expected.

(You should be able to convert the callback into a `suspendCoroutine` -- Roman gives and example of doing that here: https://stackoverflow.com/questions/48552925/existing-3-function-callback-to-kotlin-coroutines/48562175#48562175 )

Leo-Neat commented:
> I added some comments, but overall I think this really most of this needs to be split into a separate file and more, smaller methods. A method for doing the DFS, a method for printing a media item, etc...
> 
> I'm also not sure why the method requires Android N+? In generally I'd really like to keep the minSdkVersion=21 (I'd actually be fine raising it to 23, for various reasons, but not 24 yet). So I'd also prefer if you could guard the limited things that actually need Android N+ with a `Build.VERSION.SDK_INT` check around those bits, rather than annotating the entire method.

I believe that I have fixed all the issues you have brought up.

nic0lette commented on mediacontroller/src/main/java/com/example/android/mediacontroller/MediaBrowseTreeSnapshot.kt:
```diff
@@ -0,0 +1,126 @@
+     * item nodes.
+     */
+    private suspend fun runDFSOnBrowseTree(mediaItems: MutableList<MediaBrowserCompat.MediaItem>, outputStream: OutputStream) {
+        val printWriter = PrintWriter(outputStream)
```
Consider whether this bit (the output part) should be wrapped in a `withContext(Dispatchers.IO)` block.

Leo-Neat replied:
Sounds good! I will merge this then update the functionality soon.

nic0lette reviewed (approved):
Thanks :)

I don't know how much work it would be (or if it would be worth it), but I could also see value in having all the file i/o together in one bit, so the code visited all the nodes and collected them, and then there was a single spot that everything was actually output.

I'm going to mark it as approved (since the only other comment was whether you think it might make sense to do the I/O bit on `Dispatchers.IO` rather than `Dispatchers.DEFAULT`), but if you did want to do those updates (either now or in a follow up), I'd be happy to review them again.

Thank you!
